### PR TITLE
Return Live Data Pointers from Checkpoint, swap onto tensors

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96433
* #96384
* #96154
* #89146
* __->__ #95020
* #94943
* #94653

When we checkpoint the state of the private pool allocator, we will need to make sure that its current live allocated blocks will get properly cleaned up when the tensors they correspond to die. Return DataPtrs for these new allocated blocks that the callee can swap onto live Tensors.

The exact api for setting the checkpoint can be manipulated after this as the cudagraph implementation is built out, but this at least shows its sufficiently general.

This should be the last PR touching cuda caching allocator necessary for new cudagraphs integration.

Differential Revision: [D43999888](https://our.internmc.facebook.com/intern/diff/D43999888)

### CODE REVIEW

File: torch/csrc/Module.cpp [hunk 1]
```diff
@@ -1560,6 +1560,15 @@ Call this whenever a new thread is created in order to propagate values from
     return at::globalContext().linalgPreferredBackend();
   });
 
+  py_module.def(
```
Comment (zdevito) [inline torch/csrc/Module.cpp:1563]: For testing, I bet there is an already existing way to go from data_ptr to tensor.

**Review by zdevito — APPROVED** [review]
Looks good. With this API I think it does make sense to have the allocator return the set of pointers it had to free as well. Then the checkpointing operation has a simple definition: obliterate state, return the new state, return the set of pointers in the old state that are now unsafe.

File: torch/csrc/Module.cpp (hunk 1 above)
Comment (eellison) [reply torch/csrc/Module.cpp:1563]: Not to my knowledge.. if you have one link me ! For one, data pointer is not a python exposed concept.

**Comment by davidberard98** [comment]
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43999888).

**Pull request closed without merging**